### ISSUE TITLE
sum returns wrong summed value

### ISSUE THREAD

**Issue by fgelcer** [scylladb/scylla#5308]
*Installation details*
Scylla version (or git commit hash): 666.development-0.20191027.8bffb800e19
Cluster size: 3

i'm running a test to cover issue #4540 , and when running sum function on an int column i got different value in cqlsh and python (cross checked with a spreadsheet just to be sure)

here the python list:
```
[-2085001128, 1224828630, -1431197950, 19311702, -198004231, 1432216132, 1453406128, 1119828828, -1615420136, 1652206407, -528961199, -665758554, -2018652574, -1442554139, 273198835, -509786071, -2033286769, -461989264, 1373592419, 1589017456, 725238715, -511778145, 1272200182, -2037623708, -2141569087, -552139697, -929696930, 1128313594, 782926648, -119387137, 333107990, 2133468446, -1760654437, 1793599436, -1513259467, -1439350735, -1029652349, 697577542, -948457632, -1280327421, 995036095, -847139299, 2021980037, -344935548, 2092887031, -1765299170, 1994265684, -2129371931, 2019262651, -1768466049, 2010698858, 641752219, 1701290120, -111262430, 1730159495, -2044055925, -1577183351, 1708687372, 1753155263, -972238908, -1580382491, -1635617395, -574254364, -71065152, 1933338439, -40232425, -796514863, 98574291, 605118032, 251965857, -1852251173, -1665730021, 885515592, 1276072314, -383498390, 1427794312, -981566994, -465406849, 155363412, -1813187119, 804294080, 1694323328, 326311682, -1482408803, -21601469, 1496905965, 1951438782, -221505830, 1223593857, 524806869, 117471502, -266901696, -1692737124, -2020443731, 1397443608, 1512073400, 1317569903, -1351665447, -2012257263, -1070480539, 2037925248, -1737563353, -1630145333, 1958783090, 287558584, -2138258683, 1922814323, 49870899, -1270219984, -52604265, -1835685725, -1554913355, -786634998, 845600630, 1498502107, -1412841433, -1998085606, 1643399540, 1802966665, -459225337, 1390176889, -745596834, 1552545364, 1128866061, -2112175087, -1320815109, 1724010634, -194360559, -392925837, 637303578, 110648429, 13808061, 1398112451, -1793811067, -1713218409, 1013430114, 1882817160, -1804147268, 1727326571, 1525024589, 1062200845, 1820506339, 69086980, -365457797, 1125128487, -1872442265, -187029956, 769687091, -133636043, 2135308689, -240511837, -848614512, -935884996, -159189207, -159175865, -1095559090, 558961982, 92997542, 1490515400, -340915721, 1988162282, 265397100, -679965261, -1994739362, -1839044899, -723426734, 1763003385, -306295626, -466812824, 587803052, -1344341487, 55537025, 226836375, -733334188, 61916518, 312916983, -142997693, 1086401687, -400564544, -1898326200, -1751299309, -1968225992, -485915615, -35075465, -688761322, -192280499, 792928058, -311651265, -1312737877, 944292291, 1488882838, -457493158, -437103021, -1539578298, 1020449638, -423570723, -270353244, 2021160839, 1462914118, 1956196005, -1162605818, -188446740, 362784237, 649981160, 1042992049, 1737490856, 506881805, -1248538056, 3761050, 308535339, -1054124627, 439734737, -769459417, -1701126309, 47327070, 1955342520, -784040879, 372442456, 10857109, 765539076, -1607883316, -1986275298, -1335330009, 1610595530, -1795861121, 1930257142, 200967898, 1695175814, 1315619109, 13501847, 373060907, -328926303, -872630877, 150113861, 462480676, 1856474537, 1658804618, -1640344893, -1154534054, -553338743, 931370601, -651071949, -650168341, -2131989450, -91559492, -537556626, 96075472, 1916767906, -1405890700, -424200248, 455013064, 606996513, -946585406, 1292698144, -1939306523, 331484099, -2007167087, 539028950, -1111350616, -1611943499, -969543420, 1534584713, 1786205795, -922115228, -800367986, 646915858, -2096718071, -272480244, -1930710021, -384362678, 1121251154, -1422849782, -313198573, -1663671330, 1378060041, 374264245, -268043478, 316278274, 1627000671, 952884459, 1116712458, -891219618, 1795174305, -391606908, 1055134928, 2094187789, -1530009839, -1171155736, -1786954196, 504727446, -2116892545, -1164978127, -1412290875, -1745869214, 664917055, -2136205236, -1297655733, 677603604, -1953473868, 365953173, -1636174878, -879913808, 161063713, -238888620, 751766635, -1260599620, -643095044, -985233858, -1936662259, 967953189, 1788359687, 2123038867, -1751831212, 854067075, 621189982, 836852774, 1370631890, -1794033690, -305383970, -1731904362, -1419092769, -1629461351, 1097066082, -967223229, -1664489125, 710052743, 134748605, 1054884236, 1686483085, 1552984983, 1815196374, 1949212713, -162119599, -1279262399, 1425185539, 766205244, 1937394027, -1224578528, -399650582, 760658553, 1820372189, -1497127919, -424488305, 192502672, -1375087050, -1710042373, -511900176, 1885394340, 1129732300, 969398662, 341144413, 56123762, -1607990246, 1537070925, 96580993, -1644759883, -203585245, 17033412, 1278850111, -1201270828, 85394134, -748941411, -1019160830, 1005283630, 208537907, 628531820, 283071967, 1091658761, 195975484, -1296968041, -29208367, -575773530, -2000889747, -1501186744, 1922868069, 1973831494, -1536391608, -614698768, -31391853, 446498356, 898177974, 1405583794, -833918555, 287042886, -115445868, 2100398138, -877207393, 254689985, -691989950, -345782087, 635147606, -317850343, 2105017955, -157377101, 191423879, 2028370845, 801660853, 300095366, -126295138, 48633502, -935730087, 1821032050, 1869918246, 663145256, -1178423179, 662406780, 1610147819, 1635594821, 1696825017, 369528631, -1332049822, -1709015891, 1590437614, -1512768470, -988430186, 1394077511, -2052095060, -1872747630, 2015587621, 435686194, -1625184178, -1673185328, 1050423010, -261011274, 158306512, -1607973487, -1907296517, 1166515420, -2023532590, 1391243953, 537111429, -1273772824, -2019685382, -945778418, -1227221418, -555180567, -1481243816, 254511704, -1118989591, 590232882, 184966676, 2102650607, -1471328796, -21509033, 1330600853, -1704598854, -321641746, -546319317, 69116433, 1397025839, -52579496, 541048581, 733683460, 437068716, 1193410478, 808889155, 1762099415, -1844415347, -1351702693, 725745282, -1718245580, -726054715, 1678461497, -1527636518, 666380090, -1965084744, 1683798484, -64483554, -2000692277, 440556878, -1455602586, -1612416417, 1206915753, 1742346795, 588337458, -1145516279, 290010716, 1570960437, -1359277935, -817111989, -1503394127, 113130338, 1555210883, 1400481728, 1584466698, 578486155, -132527096, 2058024860, 733871074, -1372879063, 792260021, 1956465753, -1309514026, 920803161, 1642201973, 534228331, -777559325, -383543847, -1635343637, -412262537, 1995675332, -1437729303, 418101597, -1777518837, 1305514658, 871823622, 489309628, -2132767466, 1928155016, -136831631, 2062073028, 684111463, -1930316944, -1552324986, -1172127308, 53785752, -1018110994, -1767111583, -862683828, -702824478, -521456217, 1025831519, 334411893, -79582562, -1120931388, 1980638806, -837902042, -1296081933, -487890847, 259371953, 1533846505, 45205512, -384586656, -1140264905, 2093466666, -1367769922, 1608747386, -1410785660, -2073658695, -1157914376, -2001666338, -1783098406, -714171038, 1252610910, -281850045, 655270710, -1600952137, 459303875, 1829026664, -1078906637, 1793566542, -454167127, -683777017, 1454564589, 1027573891, -1069141171, 482686741, 1288842781, -212373457, 1581501318, -623239304, 2056707058, 267984429, 32990169, -1519460135, -1613544861, -1576085076, 213510867, 1476512085, 1319098635, -559497307, 1134370026, 143929464, 1540214511, -59577967, 1929303871, 2060745811, 1168827821, -830008641, 263627856, 1660649068, 930988336, 1031692119, -1039795039, 1484861938, 504267413, 1760365498, 1273827210, 395219590, -1566662996, -1337796887, 1940395932, -660265078, -1429107326, -628415125, 1875084283, 584725929, -848584914, 387790811, -2035636209]
```

here the query that lists the int column values and the sum query:
```
cqlsh:test_paging_size> SELECT someint from paging_test where mybool = true ;

 someint
-------------
 -1168990563
  1922093209
  1424040879
 -1751507602
 -2131003348
 -1907770961
 -1365147887
  -619988819
 -1241544812
  -738363767
   356006661
  1991768319
   929954274
   294777117
  1555200784
   235824159
  -853250457
   382466793
  1949142702
  -754452235
   475771344
  2124755509
    20273929
  -910852716
 -1387962457
 -1883279911
   630685215
 -1553615631
   206438640
  -113354207
  1649137508
  1780196226
 -1429651899
   104550572
   115692323
 -1329285898
 -1596325737
  -326998848
 -1358287642
  2032038672
    63517689
  1750281668
  -433028003
  1911792415
  -982710337
 -2099428342
  2133062984
 -2128506663
   264321678
  1795283623
 -1128361392
 -1526886494
  1841963884
 -1421772324
   836400095
 -1940835872
   761925348
 -1063705457
  -834856881
    82513080
   286093181
 -1184825185
   416678117
  1233593735
   152322262
  1609909752
 -1357476586
 -1730274913
   259628339
 -1900501450
  1903392127
  -354599730
 -1339306460
 -1338934078
 -1874810005
 -1281330519
  -228039129
 -1535103614
 -1176822882
  -960883418
  -556113919
   111487773
  -872205808
  1821753853
  -113369487
  -302326575
  1304155328
  1747270138
   307502322
  2018928500
 -1068033048
  -551689694
   701237958
  -968098074
   295718561
    78589059
   166870170
 -1954925173
  1755866419
   812496269
  1880660322
   634702706
  -848887997
  1888319166
  1918193692
   292647083
  1689588261
  1842295911
  -114480817
   230275639
  1144728230
   756622236
   733272234
  1537908203
   728166343
   377114621
 -2027926869
  1714032493
   -73746820
  1384041124
  -102271450
  1134188103
 -1998805670
 -1851425284
 -1759858913
 -1895669968
 -1044895108
  1780728106
 -1881737369
  -936257560
  1227725533
  1527077834
  -344485896
   138259717
 -1545429957
  -713133490
   602443955
 -1739368634
 -1967921445
 -1340847287
  -922055011
   312040488
 -1993208249
  1538233314
   419829504
  1847917784
 -1622538354
  -883928224
 -1904983789
  1229564384
   715533185
  1669644128
   436396536
  -461236092
   785564035
 -2101322231
  1770209428
  -891366779
   -45833305
  -398870896
 -2004572746
   591139433
  1405294569
  -251486517
   949158915
   223880285
    55480680
   669681396
  1245315997
   465057642
 -1011095065
  1708032359
    62123946
    13977756
 -1718923842
   365619978
  2036811106
 -1957934509
  -162388018
   470255804
  -679117510
  1126150584
  1571771760
   696778669
   224052267
 -2032409539
 -1177303232
   595755453
   239070397
 -1680591883
 -1951286348
   366791695
 -1561079771
   334810583
    78681212
   126722260
  -813851443
 -2068773115
   -49635259
   675535373
 -1280781403
 -1988786600
  1769638570
  1651028861
  -849781197
  -448335956
   -17426186
  1835955374
  -837262904
  -756098354
   739271359
  1419025348
  1631880072
  -713584067
  1954387457
  -408955925
 -2066332519
 -2139223406
 -1798474019
  1741649145
   256296624
  1932364848
 -1993409757
   486065138
  1020189069
 -1830474434
  -577354952
  1576786014
  1775178735
   969694279
  2146651062
  -621706900
 -1227864982
  2025002755
 -1575551432
  1067638576
  1377108438
    14991984
  -216478355
  1507720096
 -1293035131
 -1007306721
  1091692446
  1597588317
  -962249446
  -220619577
  1317942393
   962856900
   -30084966
  2058206741
 -1392160645
  -670934717
  -368795257
  1761767522
 -1563299281
  1219019380
  -366181330
 -1689665102
   774683022
  1199229584
   380108391
 -1476707529
  1526290648
  -102856896
 -1061537937
   814315031
  1560161178
   450586613
  1904308588
  1670790566
   154724801
  -390946013
  2021040882
  -884888490
   -73111892
 -1580242684
  -429520983
  1812418899
 -1369589451
  1049101667
  -353389722
  -630969075
   829132120
  -886848481
  1452049828
  -692870115
  2035247856
  1406526812
   708307682
  -449889665
 -1586154191
  1761065768
   -11529056
  1342835481
   -57202272
   254985982
 -1064152174
 -1396925593
 -1036474804
  1144203296
  -485891120
  2105569157
 -1260890468
  -580132554
  -850599438
  1846928402
   894331606
 -1510633236
  1436272942
  1006154836
  1304109639
  -611975410
   234647330
 -1009175611
  -232081072
   380635357
  1287995606
  -390803891
  1044112664
  1710933538
  1922910572
  1721893682
 -1180470984
 -2067874757
   713234401
  -213877748
    54427120
 -1981321769
 -1221222390
 -1774580385
 -1133777132
  1921676638
   -67883584
 -1338678291
 -2010558283
   907165946
  1258546180
 -2145662381
 -1511304663
 -1944762689
 -1763095579
   942919839
  -468762075
  2087714152
  1760177114
 -1845953809
 -1414087820
   937314033
  1366124716
   427478847
 -1583849281
   281072316
  1145363546
   196441226
  -914541569
  2086496340
  1800002659
  1016850585
  -182130754
  -854258593
  1685489360
  1782732892
  1666848381
 -1822885983
  1501463021
 -1829372606
  1243555064
  -270358652
  1841108752
   216662574
   232242432
  1366070791
   500732369
 -1742073000
   426573732
  -838336131
  1788024621
   833578776
 -1957217021
  1366501097
  1131334492
  -503447613
  2088664588
  -114986181
  1772053692
   866944386
  -210637957
 -1424059960
 -1718223106
 -1221087461
 -1630088234
   908603389
 -2126837551
   914440195
   409095291
   300274456
 -2136301024
    28203231
  -593287218
  2098659690
 -1200900137
  -707331499
  -911426972
   206234238
  1885558052
 -1056130253
 -2076806401
  1956792329
 -1473992796
 -1712283000
 -1072200682
 -1238079754
  -564674570
  -749973740
  -238243570
   315748251
   279074796
    30159467
   985553281
  1744648112
   162919574
  1312003037
    85513266
   745858239
   309827014
   392638693
  -860873714
  1565296784
 -1738598229
   -99686166
 -1147261564
  -904951299
 -1812030394
  1084263525
  -398107775
 -1364332333
  1513510364
 -1041076894
 -1083760437
  1394131313
 -1830727046
  2052351503
  -185571601
  1351866448
 -1062719374
  -159898658
   766740992
  -951727336
   122126395
 -1528641559
  1899720844
 -1686244363
  -662387956
 -1388528894
  -806253740
 -1272254009
 -1730715733
  -763201875
  1576801809
  1221377873
   591994849
   704349977
  2084294189
 -1476351902
  2023015858
   454735896
 -1580575534
 -1099085601
  -409898387
  -237810682
 -1236036327
   906596477
 -1401103479
   159868534
 -1539749034
 -1599760832
  -250636515
  1126867921
 -1575837048
  1204482645
 -1541941969
  -657647551
   626759512
   484108857
  -328906392
  -112674335
  1246465725
   380640885
 -1368098400
   781970679
   921937991
   -59940888
   937294093
  -313243922
  -839871414
   -71130459
 -1308496639
 -1849987658
 -1712445651
  -723706836
   484003569
  1072269154
     5393075
  -754537078
   677254920
  -768956280
  -654211146
  1142189298
    42226281
 -1095199275
 -1493480761
 -2007063393
  1871109382
 -1236788133
   490448207
 -1612143557
  1202448603
     -790849
   983526042
   145863766
   451073381
   159558465
   158959599
  -835436736
   298974256
  -868009577
 -1287687639
   830608507
 -2012254996
  -243606169
  -556484773
 -2066306165
  -209674237
 -1992711952
  2085342045
  -867042977
   223855686
   367716580
   290020256
 -1674511801
   171215042
  1968787467
   967186378
  1734819066
  1762701088
 -1012863787
 -1041739881
  -599524732
 -1621455941
 -1144157257
   388504461
   -25117067
  2116992479
  -113210381
   646079226
  2095245555
  1420512445
 -1533831427
 -1190965080
  1177035751
 -1868084135
  -861579868
   789550590
  1859321051
 -1465805184
 -1582638280
 -1634191499
 -2051865079
 -1283208014
  1188250721
 -1518870900
   278842987
  -893136660
   812059192
  1455530778
  -840176910
  1244176514
 -1469665118
  1054778820
 -1536011185
  -765614218
 -1125350948
  -126271912
   853865677
 -1680831584
 -1173803726
   935412103
   750197758
  -700248266
 -1260129281
   469165635
 -1765245674
 -1409048462
  1715035598
  2142336462
 -1274692367
 -1994852643
  1543580158
 -1084213588
 -1137576326

(600 rows)
cqlsh:test_paging_size> SELECT sum(someint) from paging_test where mybool = true ;

 system.sum(someint)
---------------------
           508438775

(1 rows)
```

the expected value (per spreadsheet and python) should be:
-25261365001

**Comment by fgelcer** [comment]
@slivne , what else should i put in here to make it easier to debug?

**Comment by fgelcer** [comment]
issue detected while working on this [PR](https://github.com/scylladb/scylla-dtest/pull/1125)

**Comment by psarna** [comment]
It looks like a regular overflow - this test case sums integers from its minimum value to maximum value, so it's easy for the counter to flip. Note the hex representation (in 2's complement) of both results:
 -25261365001 == FFFFFFFA**1E4E28F7**
508438775 == **1E4E28F7**

It's only an issue with our aggregation code if Cassandra automatically changes the type of `sum(int)` into `long` - that would need to be checked. Otherwise, it's just the test case that should be amended. To be absolutely sure of my statements I rerun your test cases, but with reduced range of integers to pick from, so that it's not possible to overflow during adding them up, and the test case started passing as expected.

**Comment by fgelcer** [comment]
> It looks like a regular overflow - this test case sums integers from its minimum value to maximum value, so it's easy for the counter to flip. Note the hex representation (in 2's complement) of both results:
> -25261365001 == FFFFFFFA**1E4E28F7**
> 508438775 == **1E4E28F7**
> 
> It's only an issue with our aggregation code if Cassandra automatically changes the type of `sum(int)` into `long` - that would need to be checked. Otherwise, it's just the test case that should be amended. To be absolutely sure of my statements I rerun your test cases, but with reduced range of integers to pick from, so that it's not possible to overflow during adding them up, and the test case started passing as expected.

interesting, so, if we will not be able to fix this issue soon (if at all), i will just reduce the range of my int column to be able not to overflow.

**Comment by psarna** [comment]
Cassandra also happily overflows on user request:
```cql
cqlsh:tracking> select * from t;

 id | mybool | someint    | sometext
----+--------+------------+----------
  4 |  False |          4 |        d
 27 |  False |     123456 |        d
 25 |  False | 2147483646 |        d


cqlsh:tracking> select sum(someint) from t;
 system.sum(someint)
---------------------
         -2147360190

(1 rows)

Warnings :
Aggregation query used without partition key
```

Interestingly, it also adds a warning that aggregation without a partition key was used (which makes it likely to both overflow and take a long time).

**Comment by fgelcer** [comment]
> Cassandra also happily overflows on user request:
> 
> ```sql
> cqlsh:tracking> select * from t;
> 
>  id | mybool | someint    | sometext
> ----+--------+------------+----------
>   4 |  False |          4 |        d
>  27 |  False |     123456 |        d
>  25 |  False | 2147483646 |        d
> 
> 
> cqlsh:tracking> select sum(someint) from t;
>  system.sum(someint)
> ---------------------
>          -2147360190
> 
> (1 rows)
> 
> Warnings :
> Aggregation query used without partition key
> ```
> 
> Interestingly, it also adds a warning that aggregation without a partition key was used (which makes it likely to both overflow and take a long time).

what do we plan to do in here? there are 3 (new) tests failing because of this issue...
we could try to decrease the range of int units in the test to have this working..
@amoskong , FYI.

**Comment by amoskong** [comment]
> > Cassandra also happily overflows on user request:
> > ```sql
> > cqlsh:tracking> select * from t;
> > 
> >  id | mybool | someint    | sometext
> > ----+--------+------------+----------
> >   4 |  False |          4 |        d
> >  27 |  False |     123456 |        d
> >  25 |  False | 2147483646 |        d
> > 
> > 
> > cqlsh:tracking> select sum(someint) from t;
> >  system.sum(someint)
> > ---------------------
> >          -2147360190
> > 
> > (1 rows)
> > 
> > Warnings :
> > Aggregation query used without partition key
> > ```
> > 
> > 
> > Interestingly, it also adds a warning that aggregation without a partition key was used (which makes it likely to both overflow and take a long time).
> 
> what do we plan to do in here? there are 3 (new) tests failing because of this issue...
> we could try to decrease the range of int units in the test to have this working..
> @amoskong , FYI.

Thanks.

**Comment by bhalevy** [comment]
@psarna wrote:
> Cassandra also happily overflows on user request:

So this is not a bug.
Though to be on the safe side, it might make sense to return an error on overflow unless flagged not to, but this will break compatibility with Cassandra.  Let's open a separate issue for that, and decide on that as a possible enhancement.

scylladb/scylla-dtest#1161 fixes the test to properly deal with 32-bit int overflow.

**Comment by bhalevy** [comment]
Fixed with scylladb/scylla-dtest@19ad09c8fdbf8c95fb2da6d507a4034148849552
